usr/src/lib/libzfs/common/libzfs_mount.c

```diff
@@ -88,6 +89,9 @@
 #include <sys/systeminfo.h>
 #define	MAXISALEN	257	/* based on sysinfo(2) man page */
 
+static int mount_tq_nthr = 512;	/* taskq threads for multi-threaded mounting */
```

Comment:
How did we determine this number? How does this change if you're not in the global zone trying to mount a lot of things and you have CPU caps that limit the amount of parallelism?

Reply:
@sebroy can you comment here? I don't see anything in the original review about this.

Reply:
I talked to Seb about this today, and it's like I was assuming.

The number 512 here isn't necessarily special (we could change it), we just want a number large enough such that we can create enough threads to saturate the storage IOPs. Since each thread performing a mount will generally be IO bound, we want to have enough threads mounting concurrently, so that we more effectively utilize the storage IOPs.

As far as CPU caps and zone considerations, I _think_ 512 should be fine. The performance consideration here is to make sure we issue enough IO, so we still want to have multiple threads issues mounts concurrently, even if we can only have a single thread running at a time. i.e. we want to be able to issue another mount, when the previous thread/mount blocks on IO.

With that said, I'm not familiar with the CPU caps available, so perhaps I'm overlooking something. Let me know if you'd like some more information, and/or if you still have concerns here.

Reply:
I suppose this could be made a tuneable through /etc/default/zfs?

Reply:
Well, it's pretty easy to go tune down the max lwps you have in a zone or similar. Though hopefully in those situations you won't have a lot of datasets. It's a bit unfortunate that the model doesn't fall back to the known single threaded bit on resource exhaustion. Hopefully we won't have people hitting this substantially.